src/java/org/apache/cassandra/tcm/transformations/AlterSchema.java

```diff
@@ -0,0 +1,280 @@
+import org.apache.cassandra.tcm.serialization.Version;
+import org.apache.cassandra.utils.JVMStabilityInspector;
+
+import static org.apache.cassandra.exceptions.ExceptionCode.ALREADY_EXISTS;
+import static org.apache.cassandra.exceptions.ExceptionCode.CONFIG_ERROR;
+import static org.apache.cassandra.exceptions.ExceptionCode.INVALID;
+import static org.apache.cassandra.exceptions.ExceptionCode.SERVER_ERROR;
+import static org.apache.cassandra.exceptions.ExceptionCode.SYNTAX_ERROR;
+
+public class AlterSchema implements Transformation
+{
+    private static final Logger logger = LoggerFactory.getLogger(AlterSchema.class);
+    public static final Serializer serializer = new Serializer();
+
+    public final SchemaTransformation schemaTransformation;
+    protected final SchemaProvider schemaProvider;
+
+    public AlterSchema(SchemaTransformation schemaTransformation,
+                       SchemaProvider schemaProvider)
+    {
+        this.schemaTransformation = schemaTransformation;
+        this.schemaProvider = schemaProvider;
+    }
+
+    @Override
+    public Kind kind()
+    {
+        return Kind.SCHEMA_CHANGE;
+    }
+
+    @Override
+    public final Result execute(ClusterMetadata prev)
+    {
+        Keyspaces newKeyspaces;
+        try
+        {
+            // Applying the schema transformation may produce client warnings. If this is being executed by a follower
+            // of the cluster metadata log, there is no client or ClientState, so warning collection is a no-op.
+            // When a DDL statement is received from an actual client, the transformation is checked for validation
+            // and warnings are captured at that point, before being submitted to the CMS.
+            // If the coordinator is a CMS member, then this method will be called as part of committing to the metadata
+            // log. In this case, there is a connected client and associated ClientState, so to avoid duplicate warnings
+            // pause capture and resume after in applying the schema change.
+            schemaTransformation.enterExecution();
+
+            // Guard against an invalid SchemaTransformation supplying a TableMetadata with a future epoch
+            newKeyspaces = schemaTransformation.apply(prev);
+            newKeyspaces.forEach(ksm -> {
+               ksm.tables.forEach(tm -> {
+                   if (tm.epoch.isAfter(prev.nextEpoch()))
+                       throw new InvalidRequestException(String.format("Invalid schema transformation. " +
+                                                                       "Resultant epoch for table metadata of %s.%s (%d) " +
+                                                                       "is greater than for cluster metadata (%d)",
+                                                                       ksm.name, tm.name, tm.epoch.getEpoch(),
+                                                                       prev.nextEpoch().getEpoch()));
```

Comment:
`InvalidRequestException` has always been used for error that were linked to a user request that was trying to do something that C* was not supporting (e.g. ordering by non-clustering columns). In the current patch, there are multiple places where we use that exception for errors on which the user has no control (e.g. some race conditions or internal errors). I believe that we need to change that because it can be really confusing from the user point of view.